router.js

```diff
@@ -57,30 +57,39 @@ Router.prototype.match = function (req) {
   return false;
 }
 
+var flatten = function (arr, ret) {
+  var ret = ret || [];
+  for (var i = 0, len = arr.length; i < len; ++i) {
+    if (Array.isArray(arr[i])) {
+      flatten(arr[i], ret);
+    } else {
+      ret.push(arr[i]);
+    }
+  }
+  return ret;
+}
+
 /**
  * Registers new route
  * @param method
  * @param path
- * @param callbacks
- * @param options
+ * @param callback
  */
-Router.prototype.add = function (method, path, callbacks, options) {
-  function flatten(arr, ret) {
-    var ret = ret || []
-      , len = arr.length;
-    for (var i = 0; i < len; ++i) {
-      if (Array.isArray(arr[i])) {
-        flatten(arr[i], ret);
-      } else {
-        ret.push(arr[i]);
-      }
-    }
-    return ret;
-  }
+Router.prototype.add = function (method, path, callback) {
+  var hasOptions = (typeof callback !== 'function' && !Array.isArray(callback));
+  var callbacks = [].slice.call(arguments, 2);
```

Comment:
slicing arguments prevents the whole function from being optimized by v8. I think that instead we should change the function to expect four arguments `(method, path, options, callbacks)` and, if `options` is a function or an array, set callbacks to options (or [options]) and then set options to an empty object.